Source/DafnyCompilers/CompilerRewriter.dfy

```diff
@@ -0,0 +1,2215 @@
+            }
+            case _ => {
+              // Impossible branch
+              assert false;
+            }
+          }
+        }
+        case (Failure(_), Failure(_)) => {
+          assert res.Failure?;
+          assert res'.Failure?;
+          assert EqInterpResultValue(res, res');
+        }
+        case _ => {
+          // Impossible branch
+          assert false;
+        }
+      }
+    }
+
+    // TODO: e and e' should be the same actually
+    lemma {:verify true}
+    EqInterp_Expr_UnaryOp_CanBeMapLifted_Lem(
+      e: Expr, e': Expr, op: UnaryOp, v: WV, v': WV)
+      requires !op.MemberSelect?
+      requires EqValue(v, v')
+      ensures EqPureInterpResultValue(InterpUnaryOp(e, op, v), InterpUnaryOp(e', op, v')) {
+      reveal InterpUnaryOp();
+      
+      var res := InterpUnaryOp(e, op, v);
+      var res' := InterpUnaryOp(e', op, v');
+
+      // We make a case disjunction on the final result so as to get
+      // information from the fact that the calls to ``Need`` succeeded.
+      // The Failure case is trivial.
+      if res.Success? {
+        match op {
+          case BVNot => {
+            assert v.BitVector?;
+            assert v'.BitVector?;
+          }
+          case BoolNot => {
+            assert v.Bool?;
+            assert v'.Bool?;
+          }
+          case SeqLength => {
+            assert v.Seq?;
+            assert v'.Seq?;
+            assert |v.sq| == |v'.sq|;
+          }
+          case SetCard => {
+            assert v.Set?;
+            assert v'.Set?;
+            assert |v.st| == |v'.st|;
+          }
+          case MultisetCard => {
+            assert v.Multiset?;
+            assert v'.Multiset?;
+            assert |v.ms| == |v'.ms|;
+          }
+          case MapCard => {
+            assert v.Map?;
+            assert v'.Map?;
+            assert |v.m| == |v'.m|;
+          }
+          case _ => {
+            // Impossible branch
+            assert false;
+          }
+        }
+      }
+      else {
+        assert EqPureInterpResultValue(res, res');
+      }
+    }
+
+    // TODO: we could split this lemma, whose proof is big (though straightforward),
+    // but it is a bit annoying to do...
+    // TODO: e and e' should be the same actually
+    lemma {:verify true}
+    EqInterp_Expr_BinaryOp_CanBeMapLifted_Lem(
+      e: Expr, e': Expr, bop: BinaryOp, v0: WV, v1: WV, v0': WV, v1': WV)
+      requires !bop.BV? && !bop.Datatypes?
+      requires EqValue(v0, v0')
+      requires EqValue(v1, v1')
+      ensures EqPureInterpResultValue(InterpBinaryOp(e, bop, v0, v1), InterpBinaryOp(e', bop, v0', v1')) {
+      reveal InterpBinaryOp();
+      
+      var res := InterpBinaryOp(e, bop, v0, v1);
+      var res' := InterpBinaryOp(e', bop, v0', v1');
+
+      // Below: for the proofs about binary operations involving collections (Set, Map...),
+      // see the Set case, which gives the general strategy.
+      match bop {
```

Comment:
I'd add a todo here to see if we can refactor this to avoid the explicit case disjunctions